ntcore/ds_client.py

```diff
@@ -1,21 +1,112 @@
+# validated: 2017-12-09 DV 5ab20bb27c97 cpp/DsClient.cpp cpp/DsClient.h
+#----------------------------------------------------------------------------
+# Copyright (c) FIRST 2017. All Rights Reserved.
+# Open Source Software - may be modified and shared by FRC teams. The code
+# must be accompanied by the FIRST BSD license file in the root directory of
+# the project.
+#----------------------------------------------------------------------------
+
+import json
+import threading
+
+from .support.compat import Condition
+from .support.safe_thread import SafeThread
+from .tcpsockets.tcp_connector import TcpConnector
 
 import logging
 logger = logging.getLogger('nt')
 
+
 class DsClient(object):
     
     def __init__(self, dispatcher, verbose=False):
-        pass
-    
+        self.m_dispatcher = dispatcher
+        self.verbose = verbose
+
+        self.m_active = False
+        self.m_owner = None  # type: SafeThread
+
+        self.m_mutex = threading.Lock()
+        self.m_cond = Condition(self.m_mutex)
+
+        self.m_port = None  # type: int
+        self.m_stream = None
+
     def start(self, port):
-        logging.warning("Dashboard client mode not currently implemented")
-    
+        with self.m_mutex:
+            self.m_port = port
+            if not self.m_active:
+                self.m_active = True
+                self.m_owner = SafeThread(target=self._thread_main, name='nt-dsclient')
+
     def stop(self):
-        pass
+        with self.m_mutex:
+            # Close the stream so the read (if any) terminates.
+            self.m_active = False
+            if self.m_stream:
+                self.m_stream.close()
+            self.m_cond.notify()
+
+    def _thread_main(self):
+        oldip = 0
+        connector = TcpConnector(verbose=False, timeout=1)
+
+        while self.m_active:
+            # wait for periodic reconnect or termination
+            with self.m_mutex:
+                self.m_cond.wait_for(lambda: not self.m_active, timeout=0.5)
+                port = self.m_port
+
+            if not self.m_active:
+                break
+
+            self.m_stream = connector.connect(('127.0.0.1', 1742))
+            if not self.m_active:
+                break
+            if not self.m_stream:
+                continue
+
+            while self.m_active and self.m_stream:
+                json_blob = self.m_stream.readline()
+                if not json_blob:
+                    # We've reached EOF.
+                    with self.m_mutex:
+                        self.m_stream.close()
+                        self.m_stream = None
+
+                if not self.m_active:
+                    break
+
+                try:
+                    obj = json.loads(json_blob.decode())
+                except (json.JSONDecodeError, UnicodeDecodeError):
+                    continue
+                try:
+                    ip = obj['robotIP']
```

Comment:
Generally shouldn't trust external input. Ensure this is an int, otherwise the thread will crash later on.

Reply:
Oh yeah, good point.